### ISSUE TITLE
Enhancement proposal: Import option for mere variable declarations

### ISSUE THREAD

**Issue by codemanyak** [fesch/Structorizer.Desktop#368]
A new import option might allow to convert variable declarations to non-executable instruction elements. Since #335 Structorizer tolerates declaration elements (in Pascal or VisualBasic syntax) without initialization. So why not import them to get more type information?
I tested it with the ANSI-C import project (part of #354), and it might work fine.
Maybe I should drop the Ananlyser warnings on declaration elements, they tend to be annoying. It's enough that the declarations don't count as initialization for the next use of the variable, I think.

**Comment by codemanyak** [comment]
Fixed in my development branch. Messages are still to be added to the locales system.

**Comment by GitMensch** [comment]
Would this give the option to generate better code on export? Given the "Messwert" example (when changing the keywords to the ones in the NSD before opening the file because of #370) and the C and JAVA export we end in the variable declaration

```java
	// TODO: Check and accomplish variable declarations: 
	// wert; 
	double varianz;
	double summe;
	// mittelwert; 
	// messwert; 
	int ix;
	// anzahl; 
```

and it would be really nice to be able to have an option for a language specific type definition.
In this case we may add manually for C `anzahl` -> `unsigned int`, `wert` and `mittelwert` -> `double`, `messwert` -> `double messwert[MAXWERTE]`.
For Java we may choose `anzahl` -> `int`, `wert` and `mittelwert` -> `double`, `messwert` -> `List<double> messwert`.

This way the generators can later be tweaked to generate different code (for example using `messwert.add`).

**Comment by codemanyak** [comment]
This is a continuous improvement and refinement process. I have already taught Structorizer to derive as much type indications as possible from literals, parameter or explicit declarations as possible. The "mittelwert" example shows the current capabilities. Some of the types were obvious, others remain comments because there is neither a literal assignemnt nor explicit type info.
Remember that since enhancement #335, you may place declarations in instruction elements. To avoid too much ambiguousness, mere declarations are only recognised if they have either Pascal or BASIC syntax, i.e. if they start with `var` or `dim`. On assignments, in contrast, you may also use C or Java style notations. Things get rather tricky with arrays, however. By now we have n opportunity at all to specify structs (records), pointers and other complex types apart from array. (Untyped arrrays together with Java references allow complex data structures nevertheless (see BinSearchTree example), but on export this is a dead end.
Structorizer tries to convert type specifications into the respective target language on export, and tries it as well on import. With an expression grammar and with a little help of our new friend, the GOLDParser, a lot more sophisticated type retrieval will be possible.
Structorizer wil always be a kind of intemediate language. The less specific you formulate the algorithm detals, the easier it gets to export and import.
A declaration `VAR number: REAL` will e.g. be translated to `double number` on export to C, C++, C#, Java etc. and might look like `dim number as double` in (modern) BASIC.
To achieve this, there are type name and structure mappings definied in the genearators. A method `transformTypes()` is to be configured by every Generator subclass for hat purpose. And the `nbuildNSD()` submethods of the Parsers may certainly also be equipped with suclassable type mapping methods for the import in order to unify too extreme differences.

**Comment by GitMensch** [comment]
I don't know how it was before/after #355 :-)
Good to know that declarations can be put into instruction elements, but this leads to code that will only compile to a specific language - or are they translated on generating the source?
Using a "define language specific" approach (maybe even a language specific header/footer for a single instruction, but I'll come back to this later) helps here: You can define them for each "target" language the NSD has and the export works. Normally you'd define them only for one language (or even better: have it defined from the source import) and generating for other languages will lead to a TODO stub (as it is now) - but if you want to do a better generation you simply add the language specific defines and now are able to export nice for two languages.
I'm really looking forward for the C import to see and check this. Then enhance the COBOL import myself with a given and likely good documented new sample of the C import.

**Comment by codemanyak** [comment]
> Good to know that declarations can be put into instruction elements, but this leads to code that will only compile to a specific language - or are they translated on generating the source?

Of course they are translated - unless you switch translation off. With weakly typed languages like Python, the declarations are simply suppressed. With languages like Pascal or ancient C, they are neatly placed to the beginning - where they belong - etc. There are some algorithms in Structorizer. Just try. Or have a look to the manual.

> Using a "define language specific" approach (maybe even a language specific header/footer for a single instruction, but I'll come back to this later) helps here

This is exactly what I would like to avoid. Ideally Structorizer finds out what you want and translates it according to the syntax conventions of the target language. Designing algorithms in Structorizer is to require as few language-specific knowledge as possible for the user. This requires of course somewhat more intelligence in the generators and parsers. This is where Structorizer aims to go.

**Comment by GitMensch** [comment]
I understand the goal "if possible use no language specific stuff" and it is a good one. And in any case it is best to convert data types on input to "most simple" datatypes and convert these back to a matching type (not necessarily the original data type) back.
But as you already said: this won't work with structures or specific objects (I'm really keen to see what happens when importing them in C).
Importing COBOL (as a complete program) will get hard at this case as you normally have structured data types, every data has a length (the "modern" COBOL has general types like `BINARY-LONG` but this isn't part of COBOL 85). Very simple sample:

```cobol
       01 cust-record.
         02 cust-id               pic 9(10).
         02 cust-name             pic x(50).
         02 cust-birthday         pic 9(08).
         02 cust-address.
           05 cust-street         pic x(50).
           05 cust-city           pic x(50).
           05 cust-code           pic 9(05).
```

How do you think you'd like to have this imported when the parser finished?

What I suggested was with following in mind:

* a NSD itself should be general, not specific for a computer language
*  during a possible import you may have a block of instructions that does a single thing, for example "read customer data" - depending on the language this can be a lot of code already
* I consider it a good practice to "move" these parts to a sub-program and have an instruction for "read customer data"
* in the sub it is still no good idea to have all those code lines in but to generalize them, for example into "open file", "read file", "close file"
* to not "loose" the code it would be nice to move it from the NSD instruction to a language specific "implementation" property

The file parts seem to be able to be simplified via the FileAPI but I still look for some nice samples of this that export to multiple languages ;-)

Some thing that is maybe a better example is the sub "create and print report line".

This way the NSD *is* clean and works - the export to the importing language is fine, too. Because your NSD is clean another human can load it and add the language specific parts for "his" language.

I see that it isn't the main goal to export to sources that compile directly to programs that work but I still think to extend the current approach (later) for making this possible can not be a bad idea.

**Comment by codemanyak** [comment]
I have to admit that I obviously don't know the first thing about COBOL. I guess your sample represents some kind of nested struct declaration, but I have no idea,

- whether hyphens are allowed as part of an identifier
- what the strange `pic something` types (?) are meant to specify
- how you get up to the parent level again
- whether the "line numbers" control the nesting depth or simply show it (in a strange way 01 - 02 - 05??)

If indentation matters then good luck... Hm well, at least for COBOL 85 there is already a suitable grammar for the GOLDParser. So we won't start from scratch. Perhaps is the semantics perceivable from the names of the syntactic rules...
I haven't had a deeper look into them.

And honestly, I have no idea how to represent something like that in Structorizer, such that it might at least partially work.
By the way struct and union definitions in C are accepted by the GOLDParser under certain conditions (they must be global), but they cannot be synthesized for Structorizer unless we provide some data construction tool (which is already some kind of internal enhancement request...).

**Comment by GitMensch** [comment]
The global thing is no problem with COBOL, it is all global to the program / function... see COBOL definitions as C structs. A short intro: Hypens are allowed, most COBOL types have a strict size, for example `PIC XXXX` (a field with 4 alphanumeric chars) which may be written as `PIC X(4)`, too - numeric fields have a concrete scale and precision, for example `PIC S9(6)V99` with `S` showing it is signed, `V` for the decimal comma/period so a field that can store any value between -999999.99 to 999999.99.

The numbers are variable levels, it starts with `01` and nests the fields. In the given sample above you can enter every data that has a `PIC`, then do `INITIALIZE cust-address` and you have the complete customer in one variable with the address cleared.
`MOVE cust-address TO bill-address` moves the complete address content - if `bill-address` has different size in the fields you have bad data in - (sub)field lengths are quite important...
I'll try to get the import to include these definitions the same way it is done with C structs.

The hardest part on the import is the structuring - you can `PERFORM` (read as NSD CALL) program parts (so called `SECTION`s or paragraphs (the first can contain multiple of the second, but you can use the second without defining an outer `SECTION`) but you can "fall through" these, too (I would simply ignore this for the first approaches), old code often has stuff like `GO TO par1 par2 par3 ... DEPENDING ON var` or `PERFORM par1 THRU par6` (execution goes to `par1` until a paragraph named `par6` is finished, they may only be in order with the `GO TO`s in them - but again: I'd ignore this for the first approach) - new code often is quite well structured (I recently searched for a `GO TO` in production code for more than a minute, "this code surely has one in...nope, but that one...nope..." to see I really have to check the disk via `grep` to find the rare occurences - and often it is used as a "jump to end of the `SECTION`" which can be written better nowadays coding `EXIT SECTION`).

But to keep the focus: let's get the C import rolling and a "howto" for changing/compiling the GOLDparsers, then start with the next language - if it is COBOL, which I hope I kick in and we'll provide some beautiful stuff :-)

**Comment by codemanyak** [comment]
Just lest it should be forgotten: The D7Parser is still to be enabled to process variable declarations (if the import option is active)...

**Comment by codemanyak** [comment]
D7Parser is now enabled to process variable declarations.

**Comment by GitMensch** [comment]
All current active parsers include the option, therefore it can be closed, couldn't it?
The only thing that seems open is:

> By the way, struct and union definitions in C are accepted by the ANSI-C grammar for GOLDParser under certain conditions (they must be global), but they cannot be synthesized for Structorizer unless we provide some data construction tool (which is already some kind of internal enhancement request...).

But this may be moved to a new issue, what do you think? In any case it includes changes in the current exporters...

It looks for nearly all COBOL data items this "struct definition" would be a needed item to fully import something reasonable (I'd just skip the data definitions for now, adding them as a comment only; something that could be a good first handling for the C structs/unions, too - if it isn't done already).

**Comment by codemanyak** [comment]
> By the way, struct and union definitions in C are accepted by the ANSI-C grammar for GOLDParser under certain conditions (they must be global),...

I had already lifted this restriction (global only) by a little grammar modification.

> It looks for nearly all COBOL data items this "struct definition" would be a needed item to fully import something reasonable (I'd just skip the data definitions for now, adding them as a comment only; something that could be a good first handling for the C structs/unions, too - if it isn't done already).

Well I had also come to the conclusion that in a first step struct definitions should be imported as comments at least but that on the long run we will have to support data structures in Structorizer..

**Comment by codemanyak** [comment]
On this occasion, I also came to the conclusion that it might be helpful to actually support the concept of constants in Structorizer (see #388).

**Comment by codemanyak** [comment]
In general this enhancement is now sufficiently prepared for release 3.27.
